fish-rust/src/fish.rs

```diff
@@ -0,0 +1,877 @@
+    if unsafe { libc::getrusage(libc::RUSAGE_SELF, rs.as_mut_ptr()) } != 0 {
+        let s = CString::new("getrusage").unwrap();
+        unsafe { libc::perror(s.as_ptr()) }
+        return;
+    }
+    let rs: libc::rusage = unsafe { rs.assume_init() };
+    let rss_kb = if cfg!(target_vendor = "apple") {
+        // mac use bytes.
+        rs.ru_maxrss / 1024
+    } else {
+        // Everyone else uses KB.
+        rs.ru_maxrss
+    };
+
+    let _ = writeln!(dest, "  rusage self:");
+    let _ = writeln!(dest, "      user time: {} ms", tv_to_msec(&rs.ru_utime));
+    let _ = writeln!(dest, "       sys time: {} ms", tv_to_msec(&rs.ru_stime));
+    let _ = writeln!(
+        dest,
+        "     total time: {} ms",
+        tv_to_msec(&rs.ru_utime) + tv_to_msec(&rs.ru_stime)
+    );
+    let _ = writeln!(dest, "        max rss: {} kb", rss_kb);
+    let _ = writeln!(dest, "        signals: {}", rs.ru_nsignals);
+}
+
+fn determine_config_directory_paths(argv0: &wstr) -> ConfigPaths {
+    // PORTING: why is this not just an associated method on ConfigPaths?
+
+    let mut paths = ConfigPaths::default();
+    let mut done = false;
+    // FIXME: osstr
+    let exec_path = get_executable_path(&argv0.to_string());
+    if let Ok(exec_path) = exec_path.canonicalize() {
+        FLOGF!(
+            config,
+            format!("exec_path: {:?}, argv[0]: {:?}", exec_path, argv0)
+        );
+        // TODO: we should determine program_name from argv0 somewhere in this file
+
+        // Detect if we're running right out of the CMAKE build directory
+        if exec_path.starts_with(OUT_DIR) {
+            let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+            FLOGF!(
+                config,
+                "Running out of target directory, using paths relative to CARGO_MANIFEST_DIR:\n",
+                manifest_dir.display()
+            );
+            done = true;
+            paths = ConfigPaths {
+                data: manifest_dir.join("share"),
+                sysconf: manifest_dir.join("etc"),
+                doc: manifest_dir.join("user_doc/html"),
+                bin: OUT_DIR.into(),
+            }
+        }
+
+        if !done {
+            // The next check is that we are in a reloctable directory tree
+            let installed_suffix = Path::new("/bin/fish");
+            let just_a_fish = Path::new("/fish");
+            let suffix = if exec_path.ends_with(installed_suffix) {
+                Some(installed_suffix)
+            } else if exec_path.ends_with(just_a_fish) {
+                FLOG!(
+                    config,
+                    "'fish' not in a 'bin/', trying paths relative to source tree"
+                );
+                Some(just_a_fish)
+            } else {
+                None
+            };
+
+            if let Some(suffix) = suffix {
+                let seems_installed = suffix == installed_suffix;
+
+                let base_path: PathBuf = {
```

Comment:
nit: I'm not sure if there is really much value in using `PathBuf` here. Possibly better to use `Vec<u8>` for easier resizing and convert only after that, like before

Reply:
I would rather make an extension-method  to `PathBuf` that we can test than change this type. If something is a path using `PathBuf` is the right type. `PathBuf` does have [`shrink_to`](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.shrink_to), so we can actually perform the resizing, but finding the correct length is not readily available, since the Cpp-code used `strlen`, and a mix of `WString` and `std::string` here, I think this is preferable.

Reply:
it's fine to keep the type. I'm surprised though why it's hard to compute `strlen`, it should be the byte length of our UTF-8 string

Reply:
Sorry, I commented too hastily. It is much easier to find the length than I thought (osstr has it). I will replace this with something closer to the original.

Also: [`nix`](https://docs.rs/nix/latest/nix/trait.NixPath.html) does supply it for `PathBuf` and `Path`, I don't think we need to use that